### PR TITLE
Change throwaway parameter to include out and add test

### LINKED ISSUE #10435: Property functions with a single `_`  as an argument return bad results
```xml
<Project>
    <Target Name="Go">
      <Warning Text="Surprise! $([System.String]::Join('_', 'asdf', 'jkl'))" />
    </Target>
</Project>
```

New:
```sh-session
❯ artifacts\bin\bootstrap\net472\MSBuild\Current\Bin\MSBuild.exe .\foo.proj -tl:false
MSBuild version 17.12.0-dev-24373-01+9f69926a2 for .NET Framework

S:\msbuild\foo.proj(18,7): warning : Surprise!
```

Expected:
```sh-session
❯ msbuild .\foo.proj -tl:false
MSBuild version 17.12.0-preview-24367-07+486dbb4a9 for .NET Framework


S:\msbuild\foo.proj(18,7): warning : Surprise! asdf_jkl
```

This is the root cause of errors in https://github.com/dotnet/sdk/pull/42258 because of https://github.com/dotnet/wpf/blob/d21766ccfcdeb05aa3feeb6a195c01dbf2e1a5fb/src/Microsoft.DotNet.Wpf/src/PresentationBuildTasks/Microsoft.WinFX.targets#L376 and was introduced by https://github.com/dotnet/msbuild/commit/9f69926a2c4347dd7bc98ca636198ece35860b37, specifically 

https://github.com/dotnet/msbuild/blob/8b03742aafd18a26a5d03cd045e4d24963886503/src/Build/Evaluation/Expander.cs#L3597

### PR DESCRIPTION
**Pull request by Forgind**
Fixes #10435

### Context
#10209 was generous with what it interpreted as a throwaway parameter: underscores were often misinterpreted as throwaway parameters. This changes it to require `out _` instead of just `_`, which should resolve the problem.

### Changes Made
Look for `out _` instead of just `_`

### Testing
Created a new unit test to distinguish between _ and out _

### Notes
Alternative to #10435

I searched github for all instances of `out _` in .props, .targets, and .*proj files and found none that seemed problematic, giving me some confidence in this change. I also looked for instances of _ in .targets files and found plenty of examples that were broken with #10209, indicating that that search would've prevented this problem.

@rainersigwald, leaving this as a draft until you've had a chance to try to think of potential problems with this approach. I was satisfied with my github search.

### CODE REVIEW

**Comment by Forgind** [comment]
@rainersigwald,

I haven't come up with any more ways that this is broken; have you?

**Review by rainersigwald — APPROVED** [review]

File: src/Build/Evaluation/Expander.cs [hunk 1]
```diff
@@ -3694,6 +3703,48 @@ private bool TryExecuteWellKnownFunctionWithPropertiesParam(IPropertyProvider<T>
...
+                    {
+                        object toReturn = null;
+                        foreach (MethodInfo method in methods)
+                        {
+                            Type t = method.GetParameters()[i].ParameterType;
+                            args[i] = t.IsValueType ? Activator.CreateInstance(t) : null;
+                            object currentReturnValue = GetMethodResult(objectInstance, methods, args, i + 1);
+                            if (currentReturnValue is not null)
+                            {
+                                if (toReturn is null)
```
Comment (JanKrivanek) [inline src/Build/Evaluation/Expander.cs:3720]: This is a bit nitpicky - but I believe we should distinguish between the return valua itself and the fact if it was set or not - as `null` can still be a valid return value and we can have a case of 2 matching functions one of which returns `null` and other non-null value.

**tl;dr;**: How about introducing `bool isReturnSet;` ?

**Review by JanKrivanek — APPROVED** [review]
Looks good!

I have 1 suggestion for slight improvement - but not a must have.

File: src/Build/Evaluation/Expander.cs (hunk 1 above)
Comment (Forgind) [reply src/Build/Evaluation/Expander.cs:3720]: I had to think about this a bit because although that's often true, MSBuild code doesn't have a 'null' type, and this goes almost directly to MSBuild. But what you're really saying is that we might discard a null return value incorrectly if the method really did just return null but was correctly invoked, and I think we could fail to distinguish between those.

I'm thinking about what it will ultimately look like when it ultimately gets back to the user. Perhaps a simpler fix to this would be changing line 3739 to end with `?? "null";` instead? How does that sound?

Comment (JanKrivanek) [reply src/Build/Evaluation/Expander.cs:3720]: My main concern is - you are using `toReturn` here as a way to find out if possibly more distinct overloads of a method - which is unsupported and failure case. But the additional overload would slip through here if it'd return `null`. So in such case there would be 2 similar methods, returning distinct returns, but code would silently ignore one of them (the one returning the null) and accept the other as if it'd be the only matching one.

Ping me if I'm making you even more confused ;-)

Comment (JanKrivanek) [reply src/Build/Evaluation/Expander.cs:3720]: Works as well
